File: archive/ai-sdk-js-main/packages/mcp-server/src/kuzu-store.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2024 SAP SE
/**
 * KùzuDB Graph Store for SAP AI SDK MCP Server.
 *
 * Maintains a property graph of AI SDK inference entities used to enrich
 * deployment-listing responses with model and orchestration context.
 *
 * Schema
 * ------
 * Node tables
 *   AiDeployment         – an AI Core deployed model endpoint
 *   AiModel              – a foundation model (GPT-4, Claude, etc.)
 *   OrchestrationScenario – an orchestration pipeline scenario
 *
 * Relationship tables
 *   USES_MODEL       – AiDeployment → AiModel
 *   RUNS_SCENARIO    – AiDeployment → OrchestrationScenario
 *   ROUTES_TO        – OrchestrationScenario → AiDeployment
 *
 * Usage
 * -----
 *   import { getKuzuStore } from './kuzu-store';
 *
 *   const store = getKuzuStore();
 *   await store.ensureSchema();
 *   await store.upsertDeployment('dep-abc', 'gpt-4', 'default', 'RUNNING');
 *   await store.upsertModel('gpt-4', 'azure-openai', 'Microsoft', 'chat,embed');
 *   await store.linkDeploymentModel('dep-abc', 'gpt-4');
 *   const ctx = await store.getDeploymentContext('dep-abc');
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

const KUZU_DB_PATH = (process.env['KUZU_DB_PATH'] ?? ':memory:').trim();

let _singleton: KuzuStore | null = null;

export function getKuzuStore(): KuzuStore {
  if (!_singleton) {
    _singleton = new KuzuStore(KUZU_DB_PATH);
  }
  return _singleton;
}

/** Reset the singleton (test use only). */
export function _resetKuzuStore(): void {
  _singleton = null;
}

export interface GraphContextEntry {
  relation: string;
  [key: string]: unknown;
}

export class KuzuStore {
  private dbPath: string;
  private db: any = null;
  private conn: any = null;
  private _available = false;
  private _schemaReady = false;

  constructor(dbPath: string = ':memory:') {
    this.dbPath = dbPath;
    this._available = this._initDb();
  }

  // ---------------------------------------------------------------------------
  // M1 – Initialisation
  // ---------------------------------------------------------------------------

  private _initDb(): boolean {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const kuzu = require('kuzu');
      this.db = new kuzu.Database(this.dbPath);
      this.conn = new kuzu.Connection(this.db);
      return true;
    } catch (err: any) {
      const msg: string = String(err?.message ?? err ?? '');
      if (msg.includes('Cannot find module') || msg.includes('MODULE_NOT_FOUND')) {
        process.stderr.write(
          "WARNING: 'kuzu' npm package not installed; graph-RAG features disabled. " +
          "Add \"kuzu\": \"^0.7.0\" to packages/mcp-server/package.json to enable.\n",
        );
      } else {
        process.stderr.write(`WARNING: KùzuDB init failed: ${msg.slice(0, 120)}\n`);
      }
      return false;
    }
  }

  available(): boolean {
    return this._available;
  }

  async ensureSchema(): Promise<void> {
    if (!this._available || this._schemaReady) return;

    const stmts = [
      // Node: an AI Core deployment
      `CREATE NODE TABLE IF NOT EXISTS AiDeployment (
        deploymentId  STRING,
        modelName     STRING,
        resourceGroup STRING,
        status        STRING,
        PRIMARY KEY (deploymentId)
      )`,
      // Node: a foundation model
      `CREATE NODE TABLE IF NOT EXISTS AiModel (
        modelId      STRING,
        modelFamily  STRING,
        provider     STRING,
        capabilities STRING,
        PRIMARY KEY (modelId)
      )`,
      // Node: an orchestration scenario
      `CREATE NODE TABLE IF NOT EXISTS OrchestrationScenario (
        scenarioId  STRING,
        name        STRING,
        description STRING,
        dataClass   STRING,
        PRIMARY KEY (scenarioId)
      )`,
      // Relationship: deployment uses a model
      `CREATE REL TABLE IF NOT EXISTS USES_MODEL (
        FROM AiDeployment TO AiModel
      )`,
      // Relationship: deployment handles a scenario
      `CREATE REL TABLE IF NOT EXISTS RUNS_SCENARIO (
        FROM AiDeployment TO OrchestrationScenario
      )`,
      // Relationship: scenario resolved to deployment
      `CREATE REL TABLE IF NOT EXISTS ROUTES_TO (
        FROM OrchestrationScenario TO AiDeployment
      )`,
    ];

    for (const stmt of stmts) {
      try {
        await this._exec(stmt);
      } catch (err: any) {
        const msg = String(err?.message ?? '');
        if (!msg.includes('already exists')) {
          process.stderr.write(`KùzuDB schema stmt skipped: ${msg.slice(0, 80)}\n`);
        }
      }
    }
    this._schemaReady = true;
  }

  // ---------------------------------------------------------------------------
  // Low-level query helper
  // ---------------------------------------------------------------------------

  private async _exec(cypher: string, params?: Record<string, unknown>): Promise<any> {
    if (params && Object.keys(params).length > 0) {
      return this.conn.execute(cypher, params);
    }
    return this.conn.execute(cypher);
  }

  // ---------------------------------------------------------------------------
  // M2 helpers – upsert operations called by kuzu_index tool
  // ---------------------------------------------------------------------------

  async upsertDeployment(
    deploymentId: string,
    modelName = '',
    resourceGroup = 'default',
    status = 'unknown',
  ): Promise<void> {
    if (!this._available) return;
    try {
      await this._exec(
        'MERGE (d:AiDeployment {deploymentId: $id}) ' +
        'SET d.modelName = $model, d.resourceGroup = $rg, d.status = $status',
        { id: deploymentId, model: modelName, rg: resourceGroup, status },
      );
    } catch {
      try {
        await this._exec(
          'CREATE (d:AiDeployment {deploymentId: $id, modelName: $model, ' +
          'resourceGroup: $rg, status: $status})',
          { id: deploymentId, model: modelName, rg: resourceGroup, status },
        );
      } catch (err2: any) {
        process.stderr.write(`upsertDeployment failed for ${deploymentId}: ${err2?.message}\n`);
      }
    }
  }

  async upsertModel(
    modelId: string,
    modelFamily = '',
    provider = '',
    capabilities = '',
  ): Promise<void> {
    if (!this._available) return;
    try {
      await this._exec(
        'MERGE (m:AiModel {modelId: $id}) ' +
        'SET m.modelFamily = $family, m.provider = $provider, m.capabilities = $caps',
        { id: modelId, family: modelFamily, provider, caps: capabilities },
      );
    } catch {
      try {
        await this._exec(
          'CREATE (m:AiModel {modelId: $id, modelFamily: $family, ' +
          'provider: $provider, capabilities: $caps})',
          { id: modelId, family: modelFamily, provider, caps: capabilities },
        );
      } catch (err2: any) {
        process.stderr.write(`upsertModel failed for ${modelId}: ${err2?.message}\n`);
      }
    }
  }

  async upsertScenario(
    scenarioId: string,
    name = '',
    description = '',
    dataClass = 'internal',
  ): Promise<void> {
    if (!this._available) return;
    try {
      await this._exec(
        'MERGE (s:OrchestrationScenario {scenarioId: $id}) ' +
        'SET s.name = $name, s.description = $desc, s.dataClass = $dc',
        { id: scenarioId, name, desc: description, dc: dataClass },
      );
    } catch {
      try {
        await this._exec(
          'CREATE (s:OrchestrationScenario {scenarioId: $id, name: $name, ' +
          'description: $desc, dataClass: $dc})',
          { id: scenarioId, name, desc: description, dc: dataClass },
        );
      } catch (err2: any) {
        process.stderr.write(`upsertScenario failed for ${scenarioId}: ${err2?.message}\n`);
      }
    }
  }

  async linkDeploymentModel(deploymentId: string, modelId: string): Promise<void> {
    if (!this._available) return;
    try {
      await this._exec(
        'MATCH (d:AiDeployment {deploymentId: $did}), (m:AiModel {modelId: $mid}) ' +
        'CREATE (d)-[:USES_MODEL]->(m)',
        { did: deploymentId, mid: modelId },
      );
    } catch (err: any) {
      process.stderr.write(`linkDeploymentModel failed ${deploymentId}->${modelId}: ${err?.message}\n`);
    }
  }

  async linkDeploymentScenario(deploymentId: string, scenarioId: string): Promise<void> {
    if (!this._available) return;
    try {
      await this._exec(
        'MATCH (d:AiDeployment {deploymentId: $did}), (s:OrchestrationScenario {scenarioId: $sid}) ' +
        'CREATE (d)-[:RUNS_SCENARIO]->(s)',
        { did: deploymentId, sid: scenarioId },
      );
    } catch (err: any) {
      process.stderr.write(`linkDeploymentScenario failed ${deploymentId}->${scenarioId}: ${err?.message}\n`);
    }
  }

  async linkScenarioDeployment(scenarioId: string, deploymentId: string): Promise<void> {
    if (!this._available) return;
    try {
      await this._exec(
        'MATCH (s:OrchestrationScenario {scenarioId: $sid}), (d:AiDeployment {deploymentId: $did}) ' +
        'CREATE (s)-[:ROUTES_TO]->(d)',
        { sid: scenarioId, did: deploymentId },
      );
    } catch (err: any) {
      process.stderr.write(`linkScenarioDeployment failed ${scenarioId}->${deploymentId}: ${err?.message}\n`);
    }
  }

  // ---------------------------------------------------------------------------
  // M3 helper – raw Cypher query used by kuzu_query tool
  // ---------------------------------------------------------------------------

  async runQuery(cypher: string, params?: Record<string, unknown>): Promise<GraphContextEntry[]> {
    if (!this._available) return [];
    try {
      const result = await this._exec(cypher, params ?? {});
      const colNames: string[] = result.getColumnNames?.() ?? [];
      const rows: GraphContextEntry[] = [];
      while (result.hasNext?.()) {
        const row: unknown[] = result.getNext();
        const entry: GraphContextEntry = { relation: '' };
        colNames.forEach((col, i) => { entry[col] = row[i]; });
        rows.push(entry);
      }
      return rows;
    } catch (err: any) {
      process.stderr.write(`KùzuDB query failed: ${err?.message}\n`);
      return [];
    }
  }

  // ---------------------------------------------------------------------------
  // M4 helpers – graph context enrichment for list_deployments
  // ---------------------------------------------------------------------------

  /** Return model and scenario context for a given deploymentId. */
  async getDeploymentContext(deploymentId: string): Promise<GraphContextEntry[]> {
    if (!this._available) return [];

    const modelRows = await this.runQuery(
      'MATCH (d:AiDeployment {deploymentId: $id})-[:USES_MODEL]->(m:AiModel) ' +
      'RETURN m.modelId AS modelId, m.modelFamily AS modelFamily, ' +
      'm.provider AS provider, m.capabilities AS capabilities, \'uses_model\' AS relation LIMIT 5',
      { id: deploymentId },
    );

    const scenarioRows = await this.runQuery(
      'MATCH (d:AiDeployment {deploymentId: $id})-[:RUNS_SCENARIO]->(s:OrchestrationScenario) ' +
      'RETURN s.scenarioId AS scenarioId, s.name AS name, ' +
      's.dataClass AS dataClass, \'runs_scenario\' AS relation LIMIT 5',
      { id: deploymentId },
    );

    return [...modelRows, ...scenarioRows];
  }

  /** Return deployments and routing context for a given scenarioId. */
  async getScenarioContext(scenarioId: string): Promise<GraphContextEntry[]> {
    if (!this._available) return [];

    const deploymentRows = await this.runQuery(
      'MATCH (s:OrchestrationScenario {scenarioId: $id})-[:ROUTES_TO]->(d:AiDeployment) ' +
      'RETURN d.deploymentId AS deploymentId, d.modelName AS modelName, ' +
      'd.status AS status, \'routes_to\' AS relation LIMIT 10',
      { id: scenarioId },
    );

    return deploymentRows;
  }
}

File: archive/ai-core-streaming/mcp_server/graph/kuzu_store.py
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2024 SAP SE
"""
KùzuDB Graph Store for AI Core Streaming MCP Server.

Maintains a property graph of streaming AI inference entities used to
enrich stream-status responses with deployment and routing context.

Schema
------
Node tables
  Deployment       – an AI Core deployed model/endpoint
  StreamSession    – an active or historical streaming session
  RoutingDecision  – a security-class routing outcome (aicore/vllm/blocked)

Relationship tables
  SERVED_BY  – StreamSession → Deployment   (session uses a deployment)
  ROUTED_AS  – StreamSession → RoutingDecision  (session's routing outcome)
  HANDLES    – Deployment → RoutingDecision  (deployment handles a route class)

Usage
-----
    from mcp_server.graph.kuzu_store import get_kuzu_store

    store = get_kuzu_store()
    store.ensure_schema()
    store.upsert_deployment("dep-abc", "gpt-4", "default", "running")
    store.upsert_stream("s1a2b3c4", "dep-abc", "active", "public")
    store.upsert_routing_decision("rd-001", "public", "aicore")
    store.link_session_deployment("s1a2b3c4", "dep-abc")
    store.link_session_routing("s1a2b3c4", "rd-001")
    ctx = store.get_stream_context("s1a2b3c4")
"""

import os
import sys
from typing import Any

KUZU_DB_PATH: str = os.environ.get("KUZU_DB_PATH", ":memory:").strip()

_singleton: "KuzuStore | None" = None


def get_kuzu_store() -> "KuzuStore":
    global _singleton
    if _singleton is None:
        _singleton = KuzuStore(KUZU_DB_PATH)
    return _singleton


def _reset_kuzu_store() -> None:
    """Reset the singleton (test use only)."""
    global _singleton
    _singleton = None


class KuzuStore:
    """Embedded KùzuDB graph store for ai-core-streaming graph-RAG features."""

    def __init__(self, db_path: str = ":memory:") -> None:
        self.db_path = db_path
        self._db: Any = None
        self._conn: Any = None
        self._available = self._init_db()
        self._schema_ready = False

    # ------------------------------------------------------------------
    # M1 – Initialisation
    # ------------------------------------------------------------------

    def _init_db(self) -> bool:
        try:
            import kuzu  # type: ignore[import]
            self._db = kuzu.Database(self.db_path)
            self._conn = kuzu.Connection(self._db)
            print(f"INFO: KùzuDB initialised at '{self.db_path}'", file=sys.stderr)
            return True
        except Exception as exc:
            msg = str(exc)
            if "No module named" in msg or "ModuleNotFoundError" in msg:
                print(
                    "WARNING: kuzu Python package not installed; graph-RAG features disabled. "
                    "Add 'kuzu>=0.7.0' to mcp_server/requirements.txt to enable.",
                    file=sys.stderr,
                )
            else:
                print(f"WARNING: KùzuDB init failed: {msg}", file=sys.stderr)
            return False

    def available(self) -> bool:
        return self._available

    def ensure_schema(self) -> None:
        if not self._available or self._schema_ready:
            return
        stmts = [
            # Node: an AI Core deployed model/endpoint
            """CREATE NODE TABLE IF NOT EXISTS Deployment (
                deployment_id  STRING,
                model_name     STRING,
                resource_group STRING,
                status         STRING,
                PRIMARY KEY (deployment_id)
            )""",
            # Node: an active or historical streaming session
            """CREATE NODE TABLE IF NOT EXISTS StreamSession (
                stream_id      STRING,
                deployment_id  STRING,
                status         STRING,
                security_class STRING,
                PRIMARY KEY (stream_id)
            )""",
            # Node: a security-class routing outcome
            """CREATE NODE TABLE IF NOT EXISTS RoutingDecision (
                decision_id    STRING,
                security_class STRING,
                route          STRING,
                PRIMARY KEY (decision_id)
            )""",
            # Relationship: session served by a deployment
            """CREATE REL TABLE IF NOT EXISTS SERVED_BY (
                FROM StreamSession TO Deployment
            )""",
            # Relationship: session routing outcome
            """CREATE REL TABLE IF NOT EXISTS ROUTED_AS (
                FROM StreamSession TO RoutingDecision
            )""",
            # Relationship: deployment handles a route class
            """CREATE REL TABLE IF NOT EXISTS HANDLES (
                FROM Deployment TO RoutingDecision
            )""",
        ]
        for stmt in stmts:
            try:
                self._exec(stmt)
            except Exception as exc:
                msg = str(exc)
                if "already exists" not in msg:
                    print(f"KùzuDB schema stmt skipped ({msg[:80]})", file=sys.stderr)
        self._schema_ready = True

    # ------------------------------------------------------------------
    # Low-level query helper
    # ------------------------------------------------------------------

    def _exec(self, cypher: str, params: dict | None = None) -> Any:
        if params:
            return self._conn.execute(cypher, params)
        return self._conn.execute(cypher)

    # ------------------------------------------------------------------
    # M2 helpers – upsert operations called by kuzu_index tool
    # ------------------------------------------------------------------

    def upsert_deployment(
        self,
        deployment_id: str,
        model_name: str = "",
        resource_group: str = "default",
        status: str = "unknown",
    ) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MERGE (d:Deployment {deployment_id: $id}) "
                "SET d.model_name = $model, d.resource_group = $rg, d.status = $status",
                {"id": deployment_id, "model": model_name, "rg": resource_group, "status": status},
            )
        except Exception:
            try:
                self._exec(
                    "CREATE (d:Deployment {deployment_id: $id, model_name: $model, "
                    "resource_group: $rg, status: $status})",
                    {"id": deployment_id, "model": model_name, "rg": resource_group, "status": status},
                )
            except Exception as exc2:
                print(f"upsert_deployment failed for {deployment_id}: {exc2}", file=sys.stderr)

    def upsert_stream(
        self,
        stream_id: str,
        deployment_id: str = "",
        status: str = "active",
        security_class: str = "public",
    ) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MERGE (s:StreamSession {stream_id: $id}) "
                "SET s.deployment_id = $dep, s.status = $status, s.security_class = $sc",
                {"id": stream_id, "dep": deployment_id, "status": status, "sc": security_class},
            )
        except Exception:
            try:
                self._exec(
                    "CREATE (s:StreamSession {stream_id: $id, deployment_id: $dep, "
                    "status: $status, security_class: $sc})",
                    {"id": stream_id, "dep": deployment_id, "status": status, "sc": security_class},
                )
            except Exception as exc2:
                print(f"upsert_stream failed for {stream_id}: {exc2}", file=sys.stderr)

    def upsert_routing_decision(
        self,
        decision_id: str,
        security_class: str = "public",
        route: str = "aicore",
    ) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MERGE (r:RoutingDecision {decision_id: $id}) "
                "SET r.security_class = $sc, r.route = $route",
                {"id": decision_id, "sc": security_class, "route": route},
            )
        except Exception:
            try:
                self._exec(
                    "CREATE (r:RoutingDecision {decision_id: $id, "
                    "security_class: $sc, route: $route})",
                    {"id": decision_id, "sc": security_class, "route": route},
                )
            except Exception as exc2:
                print(f"upsert_routing_decision failed for {decision_id}: {exc2}", file=sys.stderr)

    def link_session_deployment(self, stream_id: str, deployment_id: str) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MATCH (s:StreamSession {stream_id: $sid}), "
                "(d:Deployment {deployment_id: $did}) "
                "CREATE (s)-[:SERVED_BY]->(d)",
                {"sid": stream_id, "did": deployment_id},
            )
        except Exception as exc:
            print(f"link_session_deployment failed {stream_id}->{deployment_id}: {exc}", file=sys.stderr)

    def link_session_routing(self, stream_id: str, decision_id: str) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MATCH (s:StreamSession {stream_id: $sid}), "
                "(r:RoutingDecision {decision_id: $rid}) "
                "CREATE (s)-[:ROUTED_AS]->(r)",
                {"sid": stream_id, "rid": decision_id},
            )
        except Exception as exc:
            print(f"link_session_routing failed {stream_id}->{decision_id}: {exc}", file=sys.stderr)

    def link_deployment_routing(self, deployment_id: str, decision_id: str) -> None:
        if not self._available:
            return
        try:
            self._exec(
                "MATCH (d:Deployment {deployment_id: $did}), "
                "(r:RoutingDecision {decision_id: $rid}) "
                "CREATE (d)-[:HANDLES]->(r)",
                {"did": deployment_id, "rid": decision_id},
            )
        except Exception as exc:
            print(f"link_deployment_routing failed {deployment_id}->{decision_id}: {exc}", file=sys.stderr)

    # ------------------------------------------------------------------
    # M3 helper – raw Cypher query used by kuzu_query tool
    # ------------------------------------------------------------------

    def run_query(self, cypher: str, params: dict | None = None) -> list[dict]:
        if not self._available:
            return []
        try:
            result = self._exec(cypher, params or {})
            col_names: list[str] = result.get_column_names() if hasattr(result, "get_column_names") else []
            rows: list[dict] = []
            while result.has_next():
                row = result.get_next()
                rows.append(dict(zip(col_names, row)))
            return rows
        except Exception as exc:
            print(f"KùzuDB query failed: {exc}", file=sys.stderr)
            return []

    # ------------------------------------------------------------------
    # M4 helpers – graph context enrichment for stream_status
    # ------------------------------------------------------------------

    def get_stream_context(self, stream_id: str) -> list[dict]:
        """Return deployment and routing context for a given stream_id."""
        if not self._available:
            return []

        dep_rows = self.run_query(
            "MATCH (s:StreamSession {stream_id: $id})-[:SERVED_BY]->(d:Deployment) "
            "RETURN d.deployment_id AS deployment_id, d.model_name AS model_name, "
            "d.status AS status, 'served_by' AS relation LIMIT 5",
            {"id": stream_id},
        )
        route_rows = self.run_query(
            "MATCH (s:StreamSession {stream_id: $id})-[:ROUTED_AS]->(r:RoutingDecision) "
            "RETURN r.decision_id AS decision_id, r.security_class AS security_class, "
            "r.route AS route, 'routed_as' AS relation LIMIT 5",
            {"id": stream_id},
        )
        return dep_rows + route_rows

    def get_deployment_context(self, deployment_id: str) -> list[dict]:
        """Return sessions and routing decisions for a given deployment_id."""
        if not self._available:
            return []

        session_rows = self.run_query(
            "MATCH (s:StreamSession)-[:SERVED_BY]->(d:Deployment {deployment_id: $id}) "
            "RETURN s.stream_id AS stream_id, s.status AS status, "
            "s.security_class AS security_class, 'session' AS relation LIMIT 10",
            {"id": deployment_id},
        )
        route_rows = self.run_query(
            "MATCH (d:Deployment {deployment_id: $id})-[:HANDLES]->(r:RoutingDecision) "
            "RETURN r.decision_id AS decision_id, r.security_class AS security_class, "
            "r.route AS route, 'handles' AS relation LIMIT 10",
            {"id": deployment_id},
        )
        return session_rows + route_rows
